---
title: 'An Introduction to Single Cell Transcriptomics'
date: 2022-11-07
permalink: /posts/2012/08/library-preparation-for-next-generation-sequencing/
tags:
  - ngs
  - scRNAseq
---
<style> body {text-align: justify} </style>

### Introduction
Biology is messy. Even deciding the right questions to ask represents a large unknown biological 'space', complicated by non-linearity and layers of gradation that challenges the deliniation of where certain biological properties emerge. When you are in doubt of the question to ask, high throughput studies are the answer. Single cell transcriptomics have scaled massively since their inception in 2008 (citation 1). One archetypical application of scRNAseq has been in identifying cellular sub-types e.g. T-cell subsets in tumour microenvironments which may be predicative of cancer prognosis' and treatment response. Whether all these scRNAseq publications and consortiums have actually advanced our biological understanding is a separate quesiton, although if one wants to get published in Nature for 'discovering' yet another 'T-cell subset' it is definitely the way to advance ones career.


There are plenty medium articles explaining the 'ideal data analysis workflow' for scRNAseq, so I will refrain (for now) on contributing to an already over-populated resource. In the interim I will think about how to write (hopefully) high quality posts and how I can differentiate myself from others. After all, my posts are mainly for myself as a record (or 'second-brain) for information that I struggled to retrieve from the internet or learned through experience. If these help others, then that is a plus. This first post was written a year ago and has stagnated in my google docs since. The following will include the key principles behind scRNAseq and assumes only the understanding of the properties of DNA and RNA (e.g. complementary base pairing and the 5'-3' directionality of DNA) and transcription (Fig. 1).

![nucleic acid properties](/images/nucleic_acid_properties.png)
*Fig.1. Nucleic acids have polarity. The 3' end terminates with a hydroxyl group (-OH) and the 5' end terminates with a negatively charged phosphate (PO4~2~). DNA is transcribed by RNA polymerase into mRNA transcripts. Because mRNA is fairly ustable, these transcripts provide a good model for measuring gene expression.* 

### The rationale for single cell over bulk methods

Bulk methods are cheaper to peform and often carried out as a pre-requisite to single cell studies in order to provide some sort of hypothesis guidance. They involve lysing hundreds of thousands - millions of cells and reading out their transcriptome (in the case of RNAseq). This averaging destroys important information on cellular heterogeneity. In my opinion, the holy grail of scRNAseq would be if it could be performed simultaneously with with genomics, proteomics and metabolomics (i.e. to link DNA --> RNA --> protein --> metabolites) on the same sample over a time course because this would provide a better model to track changes in cell dynamics. 


